internal/framework/base/base.go

```diff
@@ -1,55 +1,76 @@
 package base
 
 import (
+	"context"
 	"fmt"
+	"os"
 
-	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
-	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
-	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
-	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
-	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
 	"github.com/hashicorp/terraform-provider-vault/internal/consts"
-	"github.com/hashicorp/terraform-provider-vault/internal/framework/validators"
+	"github.com/hashicorp/terraform-provider-vault/internal/provider"
 )
 
-// BaseModel describes common fields for all of the Terraform resource data models
-//
-// Ideally this struct would be imbedded into all Resources and DataSources.
-// However, the Terraform Plugin Framework doesn't support unmarshalling nested
-// structs. See https://github.com/hashicorp/terraform-plugin-framework/issues/242
-// So for now, we must duplicate all fields.
-type BaseModel struct {
-	ID        types.String `tfsdk:"id"`
-	Namespace types.String `tfsdk:"namespace"`
+type withMeta struct {
+	meta *provider.ProviderMeta
+}
+
+func (w *withMeta) Meta() *provider.ProviderMeta {
+	return w.meta
 }
 
-func baseSchema() map[string]schema.Attribute {
-	return map[string]schema.Attribute{
-		// Required for acceptance testing
-		// https://developer.hashicorp.com/terraform/plugin/framework/acctests#no-id-found-in-attributes
-		"id": schema.StringAttribute{
-			Computed:            true,
-			MarkdownDescription: "ID required by the testing framework",
-		},
-		consts.FieldNamespace: schema.StringAttribute{
-			Optional: true,
-			PlanModifiers: []planmodifier.String{
-				stringplanmodifier.RequiresReplace(),
-			},
-			MarkdownDescription: "Target namespace. (requires Enterprise)",
-			Validators: []validator.String{
-				validators.PathValidator(),
-			},
-		},
+// ResourceWithConfigure is a structure to be embedded within a Resource that
+// implements the ResourceWithConfigure interface.
+type ResourceWithConfigure struct {
+	withMeta
+}
+
+// Configure enables provider-level data or clients to be set in the
+// provider-defined Resource type.
+func (r *ResourceWithConfigure) Configure(_ context.Context, request resource.ConfigureRequest, response *resource.ConfigureResponse) {
+	if v, ok := request.ProviderData.(*provider.ProviderMeta); ok {
+		r.meta = v
```

Comment:
should this log or return any errors if we are unable to get the `ProviderMeta` from the request Data?

Reply:
I don't think it is required. In that case we would already get an error from the Provider's Configure method here: https://github.com/hashicorp/terraform-provider-vault/blob/df49fbf6c4b360f29162a6adc5ce7c01f65b3bc5/internal/provider/fwprovider/provider.go#L174-L181

which is where request.ProviderData comes from.

Reply:
Ah gotcha, thanks for the clarification!